internal/rpc/server/server.go

```diff
@@ -0,0 +1,129 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/aquasecurity/trivy/pkg/utils"
+
+	"github.com/aquasecurity/trivy-db/pkg/db"
+
+	"golang.org/x/xerrors"
+
+	"github.com/aquasecurity/trivy/internal/server/config"
+
+	dbFile "github.com/aquasecurity/trivy/pkg/db"
+	rpc "github.com/aquasecurity/trivy/rpc/detector"
+
+	"github.com/aquasecurity/trivy/internal/rpc/server/ospkg"
+	"github.com/aquasecurity/trivy/pkg/log"
+	"github.com/twitchtv/twirp"
+)
+
+func ListenAndServe(addr string, c config.Config) error {
+	requestWg := &sync.WaitGroup{}
+	dbUpdateWg := &sync.WaitGroup{}
+
+	withWaitGroup := func(base http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Stop processing requests during DB update
+			dbUpdateWg.Wait()
+
+			// Wait for all requests to be processed before DB update
+			requestWg.Add(1)
+			defer requestWg.Done()
+
+			base.ServeHTTP(w, r)
+
+		})
+	}
+
+	go updateDBWorker(c.AppVersion, c.CacheDir, dbUpdateWg, requestWg)
+
+	mux := http.NewServeMux()
+
+	osHandler := rpc.NewOSDetectorServer(&ospkg.Server{}, nil)
+	mux.Handle(rpc.OSDetectorPathPrefix, withToken(withWaitGroup(osHandler), c.Token))
+
+	libHandler := rpc.NewLibDetectorServer(initializeLibServer(), nil)
+	mux.Handle(rpc.LibDetectorPathPrefix, withToken(withWaitGroup(libHandler), c.Token))
+
+	log.Logger.Infof("Listening %s...", addr)
+
+	return http.ListenAndServe(addr, mux)
+}
+
+func withToken(base http.Handler, token string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+
+		if token != "" && token != r.Header.Get("Trivy-Token") {
+			rpc.WriteError(w, twirp.NewError(twirp.Unauthenticated, "invalid token"))
+			return
+		}
+		base.ServeHTTP(w, r)
+
+		end := time.Now()
+		fmt.Printf("%s\n", end.Sub(start))
+	})
+}
+
+func updateDBWorker(appVersion, cacheDir string, dbUpdateWg, requestWg *sync.WaitGroup) {
+	client := dbFile.NewClient()
+	ctx := context.Background()
+
+	for {
+		time.Sleep(10 * time.Second)
+		needsUpdate, err := client.NeedsUpdate(ctx, appVersion, false, false)
+		if err != nil {
+			log.Logger.Error(err)
```

Comment:
Add some context message why it failed unless err contains all the details to facilitate troubleshooting.

Reply:
I fixed it. 52ebb2c